src/lm/mikolov-rnnlm-lib.h

```diff
@@ -0,0 +1,361 @@
+};
+
+struct vocab_word {
+  int cn;
+  char word[MAX_STRING];
+
+  real prob;
+  int class_index;
+};
+
+const unsigned int PRIMES[] = {108641969, 116049371, 125925907, 133333309,
+  145678979, 175308587, 197530793, 234567803, 251851741, 264197411,
+  330864029, 399999781,
+  407407183, 459258997, 479012069, 545678687, 560493491, 607407037, 629629243,
+  656789717, 716048933, 718518067, 725925469, 733332871, 753085943, 755555077,
+  782715551, 790122953, 812345159, 814814293, 893826581, 923456189, 940740127,
+  953085797, 985184539, 990122807};
+const unsigned int PRIMES_SIZE  =  sizeof(PRIMES) / sizeof(PRIMES[0]);
+
+const int MAX_NGRAM_ORDER = 20;
+
+enum FileTypeEnum {TEXT, BINARY, COMPRESSED}; //  COMPRESSED not yet implemented
+
+class CRnnLM{
+ protected:
+  char train_file[MAX_FILENAME_STRING];
+  char valid_file[MAX_FILENAME_STRING];
+  char test_file[MAX_FILENAME_STRING];
+  char rnnlm_file[MAX_FILENAME_STRING];
+  char lmprob_file[MAX_FILENAME_STRING];
+
+  int rand_seed;
+  int version;
+  int filetype;
+
+  int use_lmprob;
+  real gradient_cutoff;
+
+  real dynamic;
+
+  real alpha;
+  real starting_alpha;
+  int alpha_divide;
+  double logp, llogp;
+  float min_improvement;
+  int iter;
+  int vocab_max_size;
+  int vocab_size;
+  int train_words;
+  int train_cur_pos;
+  int counter;
+
+  int anti_k;
+
+  real beta;
+
+  int class_size;
+  int **class_words;
+  int *class_cn;
+  int *class_max_cn;
+  int old_classes;
+
+  struct vocab_word *vocab;
+  void sortVocab();
+  int *vocab_hash;
+  int vocab_hash_size;
+
+  int layer0_size;
+  int layer1_size;
+  int layerc_size;
+  int layer2_size;
+
+  long long direct_size;
+  int direct_order;
+  int history[MAX_NGRAM_ORDER];
+
+  int bptt;
+  int bptt_block;
+  int *bptt_history;
+  neuron *bptt_hidden;
+  struct synapse *bptt_syn0;
+
+  int gen;
+
+  int independent;
+
+  struct neuron *neu0;		// neurons in input layer
+  struct neuron *neu1;		// neurons in hidden layer
+  struct neuron *neuc;		// neurons in hidden layer
+  struct neuron *neu2;		// neurons in output layer
+
+  struct synapse *syn0;		// weights between input and hidden layer
+  struct synapse *syn1;		// weights between hidden and output layer
+                            // (or hidden and compression if compression>0)
+  struct synapse *sync;		// weights between hidden and compression layer
+  direct_t *syn_d;		// direct parameters between input and output layer
+                        // (similar to Maximum Entropy model parameters)
+
+  // backup used in training:
+  struct neuron *neu0b;
+  struct neuron *neu1b;
+  struct neuron *neucb;
+  struct neuron *neu2b;
+
+  struct synapse *syn0b;
+  struct synapse *syn1b;
+  struct synapse *syncb;
+  direct_t *syn_db;
+
+  // backup used in n-bset rescoring:
+  struct neuron *neu1b2;
+
+  unordered_map<std::string, float> unk_penalty;
+  std::string unk_sym;
+
+ public:
+
+  int alpha_set, train_file_set;
+
+  CRnnLM()		// constructor initializes variables
```

Comment:
This constructor code should go in the .cc file since it is so huge. 
Ditto for any other long functions that appear in the header.
I won't ask you to change all the variable names though.